File: python_solutions/418.sentence-screen-fitting.cpp
// Containers
#include <deque>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#include "conf.d/say.h"

using namespace std;
int dirs[5] = {-1, 0, 1, 0, -1};

// --------------------------------------------------------
class Solution {
public:
  int wordsTyping(vector<string> &sentence, int rows, int cols) {
    int cnt = 0, n = sentence.size();
    std::unordered_map<int, int> cc;
    for (int i = 0; i < rows; i++) {
      int start = cnt % n;
      if (cc.count(start) == 0) {
        int j = 0, k = start, c = 0;
        while (j + sentence[k].size() <= cols) {
          c += 1;
          j += sentence[k].size() + 1;
          k = (k + 1) % n;
        }
        cnt += c;
        cc.emplace(start, c);
      } else
        cnt += cc[start];
    }

    return cnt / n;
  }
};

int main() {
  Solution s;
  vector<string> ss = {"a", "bcd", "e"};
  ss = {"pwievhoi",   "nngx",   "bbfnxbtx",   "qvfpise",  "xjzue",
        "ascrcacnmo", "bknodn", "spvlvcktbw", "rhklcmpd", "lwjxnfhx"};
  int r = 9593, c = 3340;
  say(s.wordsTyping(ss, r, c));
}

File: python_solutions/418.sentence-screen-fitting.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007


# https://leetcode.com/problems/sentence-screen-fitting/description/
# Medium
class Solution:
    def wordsTyping(self, sentence: List[str], rows: int, cols: int) -> int:
        cc = {}
        res, n = 0, len(sentence)
        for i in range(rows):
            start = res % n
            if start in cc:
                res += cc[start]
            else:
                cnt, j, i = 0, 0, start
                while j + len(sentence[i]) <= cols:
                    j += len(sentence[i]) + 1
                    i = (i + 1) % n
                    cnt += 1
                cc[start] = cnt
                res += cnt
        return res // n


ss = [
    "pwievhoi", "nngx", "bbfnxbtx", "qvfpise", "xjzue", "ascrcacnmo", "bknodn",
    "spvlvcktbw", "rhklcmpd", "lwjxnfhx"
]
r, c = 9593, 3340

ss, r, c = ["I", "had", "apple", "pie"], 4, 5
print(Solution().wordsTyping(ss, r, c))
